internal/vmio/vmio.go

```diff
@@ -0,0 +1,56 @@
+// Copyright 2018-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package vmio
+
+import (
+	"context"
+	"io"
+
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+)
+
+// IOConnector is function that begins initializing an IO connection (i.e.
+// vsock, FIFO, etc.), returning a channel on which the ReadWriteCloser for
+// the connection will be sent if the initialization completes successfully.
+//
+// It's provided an errG in which any asynchronous setup can be done. If
+// any setup needs to happen synchronously (before startStdioProxy in the vm
+// package returns), it can just be written outside a call to errG.Go.
+//
+// If any error occurs (in synchronous or async code), the function should
+// return it within a call to errG.Go to ensure its propogated to the caller
+// successfully.
+type IOConnector func(
+	reqCtx context.Context,
+	shimCtx context.Context,
+	logger *logrus.Entry,
+	errG *errgroup.Group,
+) <-chan io.ReadWriteCloser
```

Comment:
While I understand the motivation here and how this code evolved over time, this looks way too complex and scary from the perspective of adding new connectors (which we're going to add). This turned into a leaky abstraction. An idiomatic way in golang is not to expose concurrency primitives in APIs, leaving this decision for API user. While it works for one implementation, it might now work for another one. Also having two context parameters is bad. I think a good example of a similar solution would be the `IO` interface from `go-runc` (https://github.com/containerd/go-runc/blob/9007c2405372fe28918845901a3276c0915689a1/io.go#L25), which faces a similar problem. It's not a call to action considering the complexity, but rather my thoughts.

Reply:
>  This turned into a leaky abstraction. An idiomatic way in golang is to not to expose concurrency primitives in APIs, leaving this decision for API user.

We are the API user here though. This is not an attempt to create a general purpose library (it's under the `internal` pkg); it's just meant to adapt other libraries (the `fifo` package, [adapted here](https://github.com/sipsma/firecracker-containerd/blob/ef8f0b7984c624a96df3fec69c2da6cb63d57ad7/internal/vmio/fifo.go#L59), and the `vsock` package, [adapted here](https://github.com/sipsma/firecracker-containerd/blob/ef8f0b7984c624a96df3fec69c2da6cb63d57ad7/internal/vmio/vsock.go#L70)) to our particular IO setup needs.

The problem being solved requires we sometimes do parts of setup synchronously, sometimes do the setup asynchronously, sometimes perhaps a mix of both. I don't know how we can do that without having some concurrency primitives in the our internal interface. Open to suggestions though of course.

> I think a good example of a similar solution would be the IO interface from go-runc

Our situation is much more complex than the RunC interface; they just need to define a set of io endpoints for a single process. The `IOConnector` interface defines a way of connecting two sets of io endpoints for independent processes that are using arbitrarily different types of IO, additionally taking into account the requirement for a mix of both synchronous and asynchronous setup. I agree it's more complicated but it's because the problem is inherently much more complicated.

What specifically makes adding more implementations of a connector complex+scary? The idea is just that you do synchronous setup as normal, do asynchronous setup inside the provided errgroup and return a channel on which the final result will be published when ready. Like I said before, the code already has multiple examples of adapting functions that know nothing about `IOConnector` to the interface.

Please don't take this as just dismissing the concerns; I'm completely open to suggestions for simplifying in a way that also maintains the requirements we have. Also If there's anything I just can document more clearly or something like that too let me know.

Reply:
> our particular IO setup needs.

This are needs of the `vsock` implementation. And this complexity leaks to the interface, so other implementations have to carry the overhead. Ideally we should expose just `ReadWriteCloser` instead.

> What specifically makes adding more implementations of a connector complex+scary? 

Specifically even an empty implementation brings a channel, adds errgroup, 2 contexts and func callback. Far more than I would expect from an entity that should expose an interface close to `ReadWriteCloser`. 

I can think of an alternative implementation for vsock (compiled in my head).

```go
type Vsock struct {
	read *io.PipeReader
}

func NewVsock(port uint32) io.ReadCloser {
	reader, writer := io.Pipe()

	go func() {
		listener, err := vsock.Listen(port)
		if err != nil {
			return err
		}

		conn, err := listener.Accept()
		if err != nil {
			return err
		}

		// close writer on timeout
		// same trick for WriteCloser
		// use ctx for cancellation

		defer conn.Close()

		err = io.Copy(writer, conn)
		if err != nil {
			return err
		}
	}()

	return &Vsock{read: reader}
}

func (v *Vsock) Read(p []byte) (n int, err error) {
	return v.read.Read(p)
}

func (v *Vsock) Close() error {
	return v.read.Close()
}

```

Reply:
The needs are not just for vsock. They are for vsock, fifo and arbitrary future implementations we may need to add, which all together require a mix of sync and async setup and read/writes. What we need isn't just another `ReadWriteCloser` implementation; we are already provided those by the various libraries we are utilizing. What we need is an interface for **initializing** arbitrary `ReadWriteCloser`s and copying between them.

The sketch you wrote out would be nice, but it doesn't deal with the real-world complications that result in the `IOConnector` interface. The major ones being:
1. We want `CreateTask` and `ExecProcess` to fail if IO fails to initialize. That means that the errors returned by `Listen` and `Accept` in the example above need to be returned to the caller. However, we also need the setup to be asynchronous. If we made all setup synchronous, we actually create deadlock as the IO initialization depends on the `Create`/`Exec` call being made and the `Create`/`Exec` call depends on IO initializing successfully.
   * This combination of a need for doing something optionally asynchronously but also returning an error to the caller is where the `errgroup` comes from. The `errgroup` also allows the caller to block until the initialization is complete.
   * Like the existing examples show, you can do all sorts of asynchronous stuff in the `errgroup` or you can just do synchronous stuff and only use the `errgroup` to report the error (with an overhead of one line of code). It's a really nice, flexible interface in that respect. I don't think it's much of a cognitive burden.
2. We should really, really try to avoid increasing the amount of IO buffering we're doing as we already need to copy from task->agent->runtime->client for stdout and stderr. With an extra `io.Pipe()` that's now grown to task->io.Pipe->agent->io.Pipe->runtime->io.Pipe->client. If we have hundreds of VMs running and tasks are pumping out large amounts of data on stdout/stderr, all that extra buffering and copying is going to start adding up in terms of CPU, latency and memory.
   * So instead of creating an extra layer on top of `ReadWriteCloser`s, we just use a channel to return one of the existing implementations from the `fifo` and `vsock` libraries. This is where the returned channel comes from.
   * This has another nice side benefit of allowing us to re-use optimizations such as [this one in the golang `io` library that check if the object provided has any other methods that could result in better performance](https://github.com/golang/go/blob/7a4d02387fa16cd2a88c30357346e5cf0ae282b1/src/io/io.go#L381-L389). Instead of trying to keep up with every interface possible by adding `Read`, `Write`, `ReadFrom`, `WriteFrom`, etc., we just re-use what already exists by not wrapping objects and using standard golang channels.

For the contexts, looking at this with a fresh pair of eyes after the weekend, I see a way to have just one.
   * Originally, I did just have the `reqCtx`, which is all you think you'd need. However, that actually resulted in occasional race conditions with the `fifo` package implementation because it uses the context provided to cancel its own internal async initialization. If that didn't happen by the time the `reqCtx` got cancelled after the request was done, then you'd get mysterious errors about read/writing from an already closed FIFO. This is why the `shimCtx` got added.
   * What I'll do instead is provide just the single `vmProc.ctx` object. I already have code that cancels that context if `Create` or `Exec` fails, so it should be an adequate replacement for both of the existing contexts. I'll update with this.

So, given the above, I can get rid of one of the contexts, which leaves a context, an errgroup for reporting IO initialization errors and a channel for returning the `ReadWriteConnector`. I think each one of these are completely justified and gives us a flexible interface that allows us to plug in arbitrary implementations in the future without needing to always go back and refactor the runtime/agent implementations to deal with new async/sync setup needs.

I'm still of course open to suggestions that further simplify things without compromising on the points above and will think about it myself too, but ultimately I just don't see how an `errgroup` (which is just a light wrapper around a `sync.WaitGroup`) and a channel are that much of burden given the complications of our requirements. They are just two simple synchronization primitives that each serve a specific need.

Reply:
> That means that the errors returned by Listen and Accept in the example above need to be returned to the caller. 

Have a look on pipe's `CloseWithError` ([link](https://github.com/golang/go/blob/530097fe60c3d9feffabf52ea910bf35af9da6fb/src/io/pipe.go#L167)). You can use to to propagate an error to the caller. So when a caller tries to Read, it can receive  an error from `Accept` or `Listen` as well.

>  avoid increasing the amount of IO buffering

`io.Pipe` doesn't perform IO buffering ([link](https://github.com/golang/go/blob/530097fe60c3d9feffabf52ea910bf35af9da6fb/src/io/pipe.go#L181)). Data is copied directly.

> What we need is an interface for initializing

I think that should be done in `NewXXX` of every concrete implementation, but not in the interface.

> If we made all setup synchronous, we actually create deadlock as the IO initialization depends on the Create/Exec call being made and the Create/Exec call depends on IO initializing successfully.

We still can make it async if necessary (see my sketch above) + add some timeout mechanism. 
Let's say we do Listen, Accept and write to pipe. Or timeout if we're not writing to the pipe within X seconds. In this case a caller tries to `Read` and receives either data or error timeout, that we send through `CloseWithError`.

Reply:
> Have a look on pipe's CloseWithError
> ...
> In this case a caller tries to Read and receives either data or error timeout, that we send through CloseWithError.

Synced up offline about this too, but I don't see how we can use `CloseWithError`. We need to check if there was an error initializing the `ReadWriteCloser` and then use it if there wasn't an error. We can't wait until the actual Read/Write happens as that call can block indefinitely in the case of success and thus block Create/Exec from succeeding indefinitely.

> io.Pipe doesn't perform IO buffering

Ah I see now, that's really cool! Thanks for linking to the implementation. It uses a bunch of channels and other synchronization internally which look like they could essentially replace the channel we are creating directly. The function signature would thus change from
`func(context.Context, *logrus.Entry, *errgroup.Group) <-chan io.ReadWriteCloser`
to
`func(context.Context, *logrus.Entry, *errgroup.Group) io.ReadWriteCloser`

However, it doesn't eliminate the need for the `errgroup` AFAICT and probably results in more code being written for each implementation of the `IOConnector` func as now each one is doing an `io.Copy` in addition to the `io.Copy` that still happens between the pipes (though I understand now that doesn't actually imply more buffering). If the only benefit is that the `io.Pipe` call is making a channel for us instead of us just making it directly, I'm not convinced it's actually worth it given the extra coding overhead, but I will give it a shot to see what it looks like in practice. Will get back on this.

> I think that should be done in NewXXX of every concrete implementation, but not in the interface

This pattern isn't being used here because we technically aren't using an golang interface per-se, it's just a function signature type definition. This is because there's no need for structs to hold any persistent data.

Reply:
I gave `io.Pipe` a try here: https://github.com/sipsma/firecracker-containerd/commit/e209fdc5a0f7ed9b695570541d2d05e8361ec4eb

The only step of any significance that was eliminated was [this wait on the `ReadWriteCloser` channel here in `proxyIO`, not in the `IOConnector` implementions](https://github.com/sipsma/firecracker-containerd/commit/e209fdc5a0f7ed9b695570541d2d05e8361ec4eb#diff-f1cc210471d7d8666b600a626cdf9f84L341).

Besides that, you still have to do all the same stuff as the previous implementation except now:
1. Each implementation of the `IOConnector` func has to take care of setting up copying for the io.Pipe. I'm not seeing that this is anything but boilerplate as we still need to do the same `io.Copy` as before in `startStdioProxy`, resulting in 3 calls to `io.Copy` for every stream in every process (previously it was just once, with none in the `IOConnector` funcs).
2. There's more complication involved in thinking about when each end of every pipe and connection should be closed (and in what order so data can't get dropped)

The downsides are not the end of the world, but I don't see that the switch to `io.Pipe` provides any benefits either. I think it would have worked out if we didn't still need the `errgroup`, but like I said before AFAICT there's nothing in the `io.Pipe` interface that can replace that without creating deadlock or race conditions.

I'm going to leave it as in the PR for now, let me know if there's any objections or other ideas.

Reply:
That's fine, I don't want this PR to be blocked. I'll play with `IO` and submit a separate PR if I manage to improve this.

Reply:
I did some more thinking about this and made two more improvements that seem to work and result in a slightly easier to understand interface:
1. I just have `IOConnector` return two channels without being passed an errgroup. The synchronization ends up being exactly the same, it just results in only having to think about channels when writing an `IOConnector` func, which is a little bit nicer.
2. I wrapped `IOConnectorSet` in an interface called `IOProxy`. The intention here was to support completely different implementations of `IOProxy` that don't even use `CopyBuffer`. My hope is this makes it easy to plug in your support for Binary logging (it can just be a different `IOProxy` implementation), but worst case scenario it should just have no effect on your efforts there as it didn't change any functionality.